### PR TITLE
new: mouse delta features

### PR DESCRIPTION
**Pull request by nextdayy**
this adds various delta-related features to the mouse, including accessors for the scroll wheel, which is extremely useful in a multiversion environment. 
These changes originate from [here](https://github.com/Polyfrost/OneConfig/blob/master/versions/src/main/java/cc/polyfrost/oneconfig/platform/impl/MousePlatformImpl.java), and have been adapted to work with UniversalCraft by me.

### CODE REVIEW

**Comment by Johni0702** [comment]
We've been trying to avoid Mixin in UniversalCraft because it makes downstream usage significantly more difficult (those mixin configs need to be registered in some way).
I'm also not particularly convinced that it's a good idea to support delta methods in the first place. For starters, they don't even do what it says in the docs ("Return the change [...] since the last time this method was called"; at least the mixin accessor definitely doesn't do this) and if it did, I'd consider that fairly bad API because only one downstream mod can ever successfully use it, all other mods will just get 0. I'd argue that in any case storing the last value yourself wherever you need it will make for more reliable and predictable code.

I'm also a bit hesitant to take code from OneConfig as they seem to be placing additional restrictions on their LGPL (I'm not even sure if that's a thing you can do; certainly wouldn't be valid for GPL but I couldn't easily figure out whether it is for LGPL), and assuming it is, then you cannot contribute their code (unless you yourself hold copyright on that specific part) to a project that's just regular LGPL.

I'm not fundamentally opposed to `Mouse.Buttons`. Though I'd remove the duplicate `Button` in the names (so it's just `Mouse.Buttons.isDown(...)`), change `Clicked` to `Down` for consistency (and because imo clicking is a single event, not an ongoing action) and remove `getButtonState` (because that looks completely redundant, it's just using C-like bools instead of real `boolean`s and I don't see why you'd ever want that).

**Comment by nxtdaydelivery** [comment]
hi, thanks for your reply (and sorry for my late response)!
I see why you don't want mixins in UniversalCraft. They are just there for the accessor, and I'm not sure if you would be opposed to using a different method to access it?

And for the delta methods, their use is just for consistency between the 1.8 delta methods. I can see your point though, as yes it is fundamentally flawed. The part of this I believe to be useful is the wheel features mainly, but there are other ways this can be handled to allow for the mouse wheel state to be accessed without using a delta.

The button state methods are just for consistency with the GLFW api. (I don't really see the point in them either)

As for licensing, as a developer of OneConfig, this specific code would be added with no licensing. This code would be subsequently removed from OneConfig if this PR is merged in any way. I only the source included it for context.

File: src/main/kotlin/gg/essential/universal/UMouse.kt [hunk 1]
```diff
@@ -27,6 +60,48 @@ object UMouse {
...
+                //#endif
+            }
+
+        /** Return the change in the mouse's Y position since the last time this method was called. */
+        @JvmStatic
+        val deltaY: Double
+            get() {
+                //#if MC>=11502
+                    //#if FORGE==1
+                    //$$ return UMinecraft.getMinecraft().mouseHelper.getXVelocity()
```
Comment (Sychic) [inline src/main/kotlin/gg/essential/universal/UMouse.kt:85]: Minor nitpick, this returns the wrong value

Comment (nxtdaydelivery) [reply src/main/kotlin/gg/essential/universal/UMouse.kt:85]: oops!

**Comment by Sychic** [comment]
Personally, I think the mouse buttons would be a good addition on their own. The rest seems like it wouldn't really work due to how it uses an accessor (though it is probably also doable via reflection)

**Comment by nextdayy** [comment]
i'm going to close this PR as it is stale and not really worth it anymore. the mouse delta wheel features, which are the main point of this PR, are poor design as pointed out and are no longer necessary in our fork anyway. thank you for your time.

**Pull request closed without merging**